packages/nx/src/native/cache/http_remote_cache.rs

```diff
@@ -0,0 +1,206 @@
+use std::{
+    env,
+    fs::{self},
+    path::Path,
+};
+
+use super::cache::CachedResult;
+use flate2::Compression;
+use reqwest::{header, Client, ClientBuilder};
+use tar::{Archive, Builder};
+use tracing::trace;
+
+#[napi]
+pub struct HttpRemoteCache {
+    client: Client,
+    url: String,
+}
+
+#[napi]
+impl HttpRemoteCache {
+    #[napi(constructor)]
+    pub fn new() -> Self {
+        let mut headers = header::HeaderMap::new();
+        let auth_token = env::var("NX_SELF_HOSTED_REMOTE_CACHE_ACCESS_TOKEN");
+        if let Ok(token) = auth_token {
+            headers.insert(
+                header::AUTHORIZATION,
+                header::HeaderValue::from_str(&format!("Basic {}", token)).unwrap(),
+            );
+        }
+
+        headers.insert(
+            header::CONTENT_TYPE,
+            header::HeaderValue::from_static("application/octet-stream"),
+        );
+
+        HttpRemoteCache {
+            client: ClientBuilder::new()
+                .default_headers(headers)
+                .build()
+                .unwrap(),
+            url: env::var("NX_SELF_HOSTED_REMOTE_CACHE_SERVER")
+                .expect("NX_REMOTE_CACHE_URL must be set"),
+        }
+    }
+
+    #[napi]
+    pub async fn retrieve(
+        &self,
+        hash: String,
+        cache_directory: String,
+    ) -> napi::Result<Option<CachedResult>> {
+        let url: String = format!("{}/v1/cache/{}", self.url, hash);
+        let response = self.client.get(&url).send().await;
+        if let Ok(resp) = response {
+            trace!("HTTP response status: {}", resp.status());
+            if resp.status().is_success() {
+                // response is an application/octet-stream containing a tarball
+                // we need to extract the tarball and return the path to the extracted files
+                Ok(Some(
+                    Self::download_and_extract_from_result(resp, cache_directory.clone()).await?,
```

Comment:
Unneeded clone here. Since the `download_and_extract_from_result` method just passes the cache_directory along into another struct, we can just move the whole value.